test/sqlancer/reducer/TestEnvironment.java

```diff
@@ -0,0 +1,127 @@
+    private final MainOptions options = new MainOptions();
+    private VirtualDBGlobalState state, newGlobalState;
+
+    public TestEnvironment() throws Exception {
+        setUpTestingEnvironment();
+    }
+
+    /**
+     *
+     * @param queries:
+     *            List of Query<?>
+     *
+     * @return String of queries that appended together with '\n' separated (no '\n' at the last line)
+     */
+    public static String getQueriesString(List<Query<?>> queries) {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < queries.size(); i++) {
+            Query<?> q = queries.get(i);
+            sb.append(q.getQueryString());
+            if (i != queries.size() - 1)
+                sb.append("\n");
+        }
+        return sb.toString();
+    }
+
+    private VirtualDBGlobalState createGlobalState() {
+        try {
+            return provider.getGlobalStateClass().getDeclaredConstructor().newInstance();
+        } catch (Exception e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void initVirtualDBProvider() {
+        try {
+            ServiceLoader<DatabaseProvider> loader = ServiceLoader.load(DatabaseProvider.class);
+            for (DatabaseProvider<?, ?, ?> provider : loader) {
+                if (provider.getDBMSName().equals(databaseName)) {
+                    this.provider = (VirtualDBProvider) provider;
+                    break;
+                }
+            }
+            if (provider == null) {
+                throw new AssertionError("testing provider not registered");
+            }
+        } catch (Exception e) {
+            throw new AssertionError(e);
+        }
+
+    }
+
+    private void setUpTestingEnvironment() throws Exception {
+        initVirtualDBProvider();
+        state = createGlobalState();
+        StateToReproduce stateToReproduce = provider.getStateToReproduce(databaseName);
+
+        state.setState(stateToReproduce);
+        state.setDatabaseName(databaseName);
+        state.setMainOptions(options);
+
+        // Main.StateLogger logger = new Main.StateLogger(databaseName, provider, options);
+        // state.setStateLogger(logger);
+
+        try (SQLConnection con = provider.createDatabase(state)) {
+            state.setConnection(con);
+            newGlobalState = createGlobalState();
+            // Main.StateLogger newLogger = new Main.StateLogger(databaseName, provider, options);
+            // newGlobalState.setStateLogger(newLogger);
+            newGlobalState.setState(stateToReproduce);
+            newGlobalState.setDatabaseName(databaseName);
+            newGlobalState.setMainOptions(options);
+        }
+    }
+
+    public void setInitialStatementsFromStrings(List<String> statements) {
+        List<Query<?>> queries = new ArrayList<>();
+        for (String s : statements) {
+            Query<?> q;
+            if (s.contains("CREATE TABLE") && !s.startsWith("EXPLAIN")) {
```

Comment:
I assume we could just always use `q = new VirtualDBQuery(s);` without the check, or is it needed somewhere?